api/access/server_test.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package dps
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow/protobuf/go/flow/access"
+
+	"github.com/optakt/flow-dps/testing/mocks"
+)
+
+func TestNewServer(t *testing.T) {
+	index := &mocks.Reader{}
+	codec := &mocks.Codec{}
+
+	s := NewServer(index, codec)
+
+	assert.NotNil(t, s)
+	assert.NotNil(t, s.codec)
+	assert.Equal(t, index, s.index)
+	assert.Equal(t, codec, s.codec)
+}
+
+func TestServer_GetLatestBlockHeader(t *testing.T) {
+	t.Run("nominal case", func(t *testing.T) {
+		index := &mocks.Reader{
+			LastFunc: func() (uint64, error) {
+				return mocks.GenericHeight, nil
+			},
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, mocks.GenericHeight, height)
+
+				return mocks.GenericHeader, nil
+			},
+		}
+
+		s := baselineServer(t)
+		s.index = index
+
+		resp, err := s.GetLatestBlockHeader(context.Background(), nil)
+
+		assert.NoError(t, err)
+
+		var gotID flow.Identifier
+		copy(gotID[:], resp.Block.Id)
+		assert.Equal(t, mocks.GenericHeader.ID(), gotID)
+		assert.Equal(t, mocks.GenericHeader.Height, resp.Block.Height)
+	})
+
+	t.Run("handles indexer error on Last", func(t *testing.T) {
+		index := &mocks.Reader{
+			LastFunc: func() (uint64, error) {
+				return 0, mocks.DummyError
+			},
+		}
```

Comment:
Here you again initialize something with reduced functionality, rather than initializing the baseline that handles both needed functions and then just breaking one. It's not really a big deal here, but I just think it's not a good habit. It's like taking a shortcut. You do that throughout the PR here.

Something like `baselineIndex` would be better imo.

Reply:
I was thinking, maybe we should have a `baselineX` in the mocks package for each of the mocks. I'll do that, and this way it will make it much easier and shorter to write those test cases.

Reply:
That could work great, depending on how it's implemented :+1: